nautobot/core/tests/test_jobs.py

```diff
@@ -606,3 +608,177 @@ def test_cleanup_object_changes(self):
+        ]
+        for cloud_service in cloud_services:
+            cloud_service.cloud_networks.add(cloud_networks[0])
+
+        cloud_service_to_ignore = cloud_services[3]
+        cloud_services_pks = [str(cloud_service.pk) for cloud_service in cloud_services[:3]]
+        job_result = create_job_result_and_run_job(
+            "nautobot.core.jobs.bulk_actions",
+            "BulkEditObjects",
+            content_type=cloud_service_ct.id,
+            edit_all=False,
+            filter_query_params={},
+            post_data={
+                "pk": cloud_services_pks,
+                "description": "Example description for bulk edit",
+                "add_cloud_networks": [str(cloud_networks[1].pk)],
+                "remove_cloud_networks": [str(cloud_networks[0].pk)],
+            },
+            username=self.user.username,
+        )
+        self._common_no_error_test_assertion(
+            CloudService,
+            job_result,
+            3,
+            description="Example description for bulk edit",
+            cloud_networks__name=cloud_networks[1].name,
+        )
+        for cloud_service in cloud_services[:3]:
+            self.assertFalse(cloud_service.cloud_networks.filter(name=cloud_networks[0].name).exists())
+
+        cloud_service_to_ignore.refresh_from_db()
+        self.assertNotEqual(cloud_service_to_ignore.description, "Example description for bulk edit")
+        self.assertTrue(cloud_service_to_ignore.cloud_networks.filter(name=cloud_networks[0].name).exists())
+        self.assertFalse(cloud_service_to_ignore.cloud_networks.filter(name=cloud_networks[1].name).exists())
+
+    def test_bulk_edit_objects_with_constrained_permission(self):
+        obj_perm = ObjectPermission(
+            name="Test permission",
+            constraints={"pk": str(uuid.uuid4())},  # Match a non-existent pk (i.e., deny all)
+            actions=["change", "view"],
+        )
+        obj_perm.save()
+        obj_perm.users.add(self.user)
+        obj_perm.object_types.add(self.role_ct)
+
+        role_to_update = Role.objects.create(name="Example Role")
+        job_result = create_job_result_and_run_job(
+            "nautobot.core.jobs.bulk_actions",
+            "BulkEditObjects",
+            content_type=self.role_ct.id,
+            edit_all=False,
+            filter_query_params={},
+            post_data={"pk": [str(role_to_update.pk)], "color": "aa1409"},
+            username=self.user.username,
+        )
+        error_log = JobLogEntry.objects.get(job_result=job_result, log_level=LogLevelChoices.LOG_ERROR)
+        self.assertIn("Form validation unsuccessful", error_log.message)
+        self.assertIn(f"{role_to_update.pk} is not one of the available choices.", error_log.message)
+        role_to_update.refresh_from_db()
+        self.assertNotEqual(role_to_update.color, "aa1409")
+
+        obj_perm.constraints = {"pk__isnull": False}  # Match a non-existent pk (i.e., allow all)
```

Comment:
Would be good to also have a test with legitimately constrained permissions, i.e. allowing update of *some* objects but not others?